File: cpp/print-in-order.cpp
1114. Print in Order
Easy

336

49

Add to List

Share
Suppose we have a class:

public class Foo {
  public void first() { print("first"); }
  public void second() { print("second"); }
  public void third() { print("third"); }
}
The same instance of Foo will be passed to three different threads. Thread A will call first(), thread B will call second(), and thread C will call third(). Design a mechanism and modify the program to ensure that second() is executed after first(), and third() is executed after second().

 

Example 1:

Input: [1,2,3]
Output: "firstsecondthird"
Explanation: There are three threads being fired asynchronously. The input [1,2,3] means thread A calls first(), thread B calls second(), and thread C calls third(). "firstsecondthird" is the correct output.
Example 2:

Input: [1,3,2]
Output: "firstsecondthird"
Explanation: The input [1,3,2] means thread A calls first(), thread B calls third(), and thread C calls second(). "firstsecondthird" is the correct output.
 

Note:

We do not know how the threads will be scheduled in the operating system, even though the numbers in the input seems to imply the ordering. The input format you see is mainly to ensure our tests' comprehensiveness.

class Foo {
public:
    Foo() {
        count=0;
    }

    void first(function<void()> printFirst) {
        unique_lock<mutex> lock(mtx);
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        count++;
        cv.notify_all();
    }

    void second(function<void()> printSecond) {
        unique_lock<mutex> lock(mtx);
        while (count!=1) cv.wait(lock);
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        count++;
        cv.notify_all();
    }

    void third(function<void()> printThird) {
        unique_lock<mutex> lock(mtx);
        while (count!=2) cv.wait(lock);
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
    }
private:
    mutex mtx;
    condition_variable cv;
    int count;
};

File: python/print-in-order.py
1. Lock
from threading import Lock

class Foo:
    def __init__(self):
        self.lock1, self.lock2=Lock(), Lock()
        self.lock1.acquire()
        self.lock2.acquire()


    def first(self, printFirst: 'Callable[[], None]') -> None:

        # printFirst() outputs "first". Do not change or remove this line.
        printFirst()
        self.lock1.release()


    def second(self, printSecond: 'Callable[[], None]') -> None:
        with self.lock1:
            # printSecond() outputs "second". Do not change or remove this line.
            printSecond()
        self.lock2.release()


    def third(self, printThird: 'Callable[[], None]') -> None:
        with self.lock2:
            # printThird() outputs "third". Do not change or remove this line.
            printThird()

2. Condition
from threading import Condition

class Foo:
    def __init__(self):
        self.cv=Condition()
        self.count=0

    def first(self, printFirst: 'Callable[[], None]') -> None:
        with self.cv:
            # printFirst() outputs "first". Do not change or remove this line.
            printFirst()
            self.count+=1
            self.cv.notify_all()


    def second(self, printSecond: 'Callable[[], None]') -> None:
        with self.cv:
            self.cv.wait_for(lambda: self.count==1)
            # printSecond() outputs "second". Do not change or remove this line.
            printSecond()
            self.count+=1
            self.cv.notify_all()


    def third(self, printThird: 'Callable[[], None]') -> None:
        with self.cv:
            self.cv.wait_for(lambda: self.count==2)
            # printThird() outputs "third". Do not change or remove this line.
            printThird()

